homeassistant/components/volumio/media_player.py

```diff
@@ -118,33 +125,109 @@ def __init__(self, name, host, port, hass):
+        self._methods = {
+            "getDeviceInfo": "pushDeviceInfo",
+            "getState":"pushState",
+            "getQueue": "pushQueue",
+            "getMultiRoomDevices":"pushMultiRoomDevices",
+            "getLibraryListing": "pushLibraryListing",
+            "getMenuItems": "pushMenuItems",
+            "getUiConfig": "pushUiConfig",
+            "getBrowseSources": "pushBrowseSources",
+            "browseLibrary": "pushBrowseLibrary",
+            "search": "pushBrowseLibrary",
+            "goTo": "pushBrowseLibrary",
+            "GetTrackInfo":"pushGetTrackInfo",
+            "addWebRadio":"pushAddWebRadio",
+            "removeWebRadio":"pushBrowseLibrary",
+            "getPlaylistContent":"pushPlaylistContent",
+            "createPlaylist":"pushCreatePlaylist",
+            "deletePlaylist":"pushListPlaylist",
+            "listPlaylist":"pushListPlaylist",
+            "addToPlaylist":"pushListPlaylist",
+            "removeFromPlaylist":"pushBrowseLibrary",
+            "playPlaylist":"pushPlayPlaylist",
+            "enqueue":"pushEnqueue",
+            "addToFavourites":"urifavourites",
+            "removeFromFavourites":"pushBrowseLibrary",
+            "playFavourites": "pushPlayFavourites",
+            "addToRadioFavourites":"pushAddToRadioFavourites",
+            "removeFromRadioFavourites":"pushRemoveFromRadioFavourites",
+            "playRadioFavourites":"pushPlayRadioFavourites",
+            "getSleep":"pushSleep"
+            
+            #TODO: There are lots more, continue at L673
+        }
+                        
     async def send_volumio_msg(self, method, params=None):
+        """Handles volumio calls"""
+        
+        state_name = await self.send(method, params) 
+        if state_name is not None:
+            return getattr(self, state_name)
+        return state_name
+
+    async def send(self, method, params=None):
         """Send message."""
-        url = f"http://{self.host}:{self.port}/api/v1/{method}/"
-
-        _LOGGER.debug("URL: %s params: %s", url, params)
-
+        _LOGGER.debug("Send, method: %s params: %s", method, params)
+        
+        def api2websocket(method, params):
+            """Transform method and params from api to websocket calls."""
+            if method == "commands":
+                method = params["cmd"]
+
+                if method in params:
+                    params = params[method]
+                elif "value" in params:
+                    params = params["value"]
+                elif "name" in params:
+                    params = params["name"]
+                else:
+                    params = params["cmd"]
+                    
+            if method == params:
+                params = None
+                    
+            return method, params
+        
+        method, params = api2websocket(method, params)
+        
+        url = f"http://{self.host}:{self.port}"
```

Comment:
This should be extracted to an external package published on PyPi

See: <https://developers.home-assistant.io/docs/creating_component_code_review/#4-communication-with-devicesservices>

Reply:
Done.